### PR TITLE
fix!: typed div and safe div semantics

### PR DESCRIPTION
**Pull request by barakalon**
Most dialects throw an error for `1 / 0`. Some return `null`.

This safely transpiles in one direction - from dialects that return null to those that raise an error.

### CODE REVIEW

File: sqlglot/expressions.py
Comment (barakalon) [inline sqlglot/expressions.py:6545]: Not actually used here - I added it when transpiling NULLIF to a CASE statement. But then I couldn't find any dialects that don't support NULLIf.

**Comment by GeorgeSittas** [comment]
Dialects that raise on division by 0:
- Bigquery
- DataBricks
- [Oracle](https://stackoverflow.com/questions/5027808/how-to-avoid-divide-by-zero-error-in-an-sql-query) (?)
- Postgres
- Presto
- Redshift
- Snowflake
- [Teradata](https://stackoverflow.com/questions/39623177/teradata-returning-division-by-zero-error-how-to-adress)
- Trino
- T-SQL

Dialects that yield `NULL` on division by 0:
- ClickHouse
- DuckDB
- Hive
- MySQL
- Spark
- Spark2
- SQLite

File: sqlglot/dialects/spark2.py
Comment (GeorgeSittas) [inline sqlglot/dialects/spark2.py:104]: Why is this `False`? I think it needs to be `False` only for databricks, see https://github.com/tobymao/sqlglot/pull/2516#issuecomment-1796481170

File: sqlglot/dialects/dialect.py
Comment (tobymao) [inline sqlglot/dialects/dialect.py:186]: what about where it returns nan?

Comment (tobymao) [reply sqlglot/dialects/dialect.py:186]: is there a difference between 1 / 0 and 1.0 / 0?

File: sqlglot/expressions.py
Comment (tobymao) [reply sqlglot/expressions.py:6551]: it's fine

**Review by tobymao — APPROVED** [review]

File: sqlglot/dialects/dialect.py
Comment (barakalon) [reply sqlglot/dialects/dialect.py:186]: 1 / 0:

- ClickHouse: ?
- DuckDB: nan
- Hive: null
- MySQL: null
- Spark: ?
- Spark2: ?
- SQLite: null

@GeorgeSittas - can you fill in the blanks for clickhouse and spark?

Comment (GeorgeSittas) [reply sqlglot/dialects/dialect.py:186]: DuckDB, ClickHouse and Spark all return `NULL` for both `1 / 0` and `1.0 / 0`

Comment (barakalon) [reply sqlglot/dialects/dialect.py:186]: You sure about that for duckdb? I might be on an old version

Comment (barakalon) [reply sqlglot/dialects/dialect.py:186]: Ah, I'm calling the `.df()` method, that's why

Comment (barakalon) [reply sqlglot/dialects/dialect.py:186]: ok, looks like nothing return nan.

**Comment by barakalon** [comment]
@GeorgeSittas I'll wait for your review before merging

**Comment by GeorgeSittas** [comment]
Thanks - doing a pass at it now

File: sqlglot/optimizer/annotate_types.py
Comment (GeorgeSittas) [inline sqlglot/optimizer/annotate_types.py:274]: This isn't correct for every dialect, e.g. postgres seems to have some coercion rules going on:

```
georgesittas=# select pg_typeof(nullif(1, 1.0));
 pg_typeof
-----------
 numeric
(1 row)

georgesittas=# select pg_typeof(nullif(1.0, 1));
 pg_typeof
-----------
 numeric
(1 row)

georgesittas=# select pg_typeof(nullif(1, 1));
 pg_typeof
-----------
 integer
(1 row)
```

Comment (GeorgeSittas) [reply sqlglot/optimizer/annotate_types.py:274]: Perhaps it's safer to do a `_maybe_coerce` on the two args of `Nullif` or something

Comment (barakalon) [reply sqlglot/optimizer/annotate_types.py:274]: Cool, seems like that might be what postgres is doing

**Review by GeorgeSittas — APPROVED** [review]
LGTM - just that one last comment re: the type inference of `exp.Nullif`.

**Pull request merged**
